### PR TITLE
ENH: Use Kahan summation to calculate groupby.sum()

### LINKED ISSUE #38778: BUG: sum vs groupby.sum errors
[data_source.zip](https://github.com/pandas-dev/pandas/files/5750979/data_source.zip)

The attached data source is a subset of a larger set. I included just enough rows to show the error (when the datasource is larger, the difference of sums grows larger too).

#### Code Sample, a copy-pastable example

```
import pandas as pd
import numpy as np
pd.options.display.float_format = '{:.5f}'.format

df = pd.read_csv("data_source.csv", index_col = 0)

df.reset_index(inplace = True)

print ("Regular sum: %s\n" % df["rul_c"].sum())

print ("Regular sum on filtered column: %s\n" % df[df["cont"] == 20]["rul_c"].sum())

print ("GroupBy sum:\n%s" % df.groupby("cont")["rul_c"].sum())

```

#### Problem description

I know about floating point math and small associated "errors" but the "cont" column has an unique, not-null value, and this means all sums above should be the same (about 30880496049.43). However the groupby.sum is quite different from the expected result .45165.

The larger the dataset, the larger the difference.

```
Regular sum: 30880496049.429993

Regular sum on filtered column: 30880496049.429993

GroupBy sum:
cont
20   30880496049.45165
Name: rul_c, dtype: float64
```

#### Expected Output
All sums should be 30880496049.429993

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 3e89b4c4b1580aa890023fc550774e63d499da25
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Romanian_Romania.1252

pandas           : 1.2.0
numpy            : 1.19.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 47.1.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.3.4
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #38778 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This simplifies the op example.

I think 
```
series = Series([1e16, 99, -5e15, -5e15])
series.sum()
```
is dispatched to numpy? The result is wrong too. ``100.0``

Mean doesn't work either. I will look through the functions to determine which need Kahan summation too and open an issue to track these.

### CODE REVIEW

**Comment by phofl** [comment]
We dispatch to numpy in case of 32bit systems, which returns the wrong result

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -51,6 +51,7 @@ Other enhancements
 - :func:`pandas.read_sql_query` now accepts a ``dtype`` argument to cast the columnar data from the SQL database based on user input (:issue:`10285`)
 - Improved integer type mapping from pandas to SQLAlchemy when using :meth:`DataFrame.to_sql` (:issue:`35076`)
 - :func:`to_numeric` now supports downcasting of nullable ``ExtensionDtype`` objects (:issue:`33013`)
+- Improve numerical stability for :meth:`DataFrameGroupBy.sum()` and :meth:`SeriesGroupBy.sum()` through using Kahan summation (:issue:`38778`)
```
Comment (arw2019) [inline doc/source/whatsnew/v1.3.0.rst:54]: no brackets after `...GroupBy.sum`?

Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst]: thx

**Review by jreback — CHANGES_REQUESTED** [review]
comments, ping on green.

note that we have been seeing some recent work on *combing* the aggregation functions for 1d and groupby, this  would be a great case here. (slightly complicated because for 1d we actually dispatch to bottleneck if installed), otherwise we use the `np.nan*` functions so have to see if this is worth it).

File: doc/source/whatsnew/v1.3.0.rst [hunk 2]
```diff
@@ -51,6 +51,7 @@ Other enhancements
 - :func:`pandas.read_sql_query` now accepts a ``dtype`` argument to cast the columnar data from the SQL database based on user input (:issue:`10285`)
 - Improved integer type mapping from pandas to SQLAlchemy when using :meth:`DataFrame.to_sql` (:issue:`35076`)
 - :func:`to_numeric` now supports downcasting of nullable ``ExtensionDtype`` objects (:issue:`33013`)
+- Improve numerical stability for :meth:`DataFrameGroupBy.sum` and :meth:`SeriesGroupBy.sum` through using Kahan summation (:issue:`38778`)
```
[the remark this answers is not in the document]
Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst]: Moved

**Review by jreback — APPROVED** [review]

**Comment by phofl** [comment]
@jreback green

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**